Add celery workers to LMS and CMS for edxapp

sampaccoud opened the pull request:
## Purpose

We need to run Celery workers to handle asynchronous tasks.

## Proposal

- [x] add Redis as a separate app (to avoid redeploying it each time we redeploy edxapp) and without blue/green (because Redis does not support several instances on the same volume...) 
- [x] add Celery workers to edxapp (one for each queue in both LMS and CMS)
- [x] remove the "selector" item in dc templates as it seems useless and is generating errors with multiple deployments per service.
- [x] add vars for celery queues and use them in DCs + settings

sampaccoud committed:
✨(edxapp) add redis app

We are going to use redis for Celery instead of RabbitMQ.
Let's add a Redis running the latest stable stretch image
of Redis. It works in OpenShift without any modification.
We declare it as a separate app so that we can exclude it
from our frequent updates of the other applications. Note
that blue/green is not activated on Redis because only
one container at a time should be mounted on the Redis
data volume.

sampaccoud committed:
✨(edxapp) add Celery workers for lms and cms

The Celery workers run on the same image as the lms and cms
but with an overriden command. We run several containers,
one for each type of queue: default, high, low (and high_mem
only for the lms).

sampaccoud committed:
🔥(dc) remove selector item from dc template files

We don't know why there is a selector item on deployment
configs. It appears to be useless and raises some errors
when we have several dc files per service.

jmaupetit commented:
In your commit message: 

> **sparkles(edxapp) add redis app**
>
> [...] Note that blue/green is note activated on Redis because only one container at a time should be mounted on the Redis data volume.

`s/note/not/`

jmaupetit commented on apps/redis/templates/app/svc.yml.j2:
```diff
@@ -0,0 +1,19 @@
+    app: redis
+    service: app
+    version: "{{ redis_app_image_tag }}"
+  name: "{{ redis_app_host }}"  # name of the service should be host name in nginx
```
Can you explain this comment about `nginx`? Do you mean in `edxapp` settings?

jmaupetit commented on apps/edxapp/templates/cms/_configs/settings.yml.j2:
```diff
@@ -6,3 +6,8 @@ SITE_NAME: "{{ project_display_name }}"
+
+# Celery Broker
+CELERY_BROKER_TRANSPORT: redis
+CELERY_BROKER_HOSTNAME: redis
```
Should be `"{{ redis_app_host }}"`?

jmaupetit commented on apps/edxapp/templates/cms/dc_worker_default_priority.yml.j2:
```diff
@@ -0,0 +1,7 @@
+          - "/bin/bash"
+          - "-c"
+          - python manage.py lms celery worker --loglevel=info --queues=edx.lms.core.high --hostname=edx.lms.core.default.%%h --concurrency=1
+{% endblock %}
```
A carriage return is missing at the end of all dc_worker* templates.

jmaupetit commented on apps/edxapp/templates/cms/dc_worker_default_priority.yml.j2:
```diff
@@ -0,0 +1,7 @@
+        command:
+          - "/bin/bash"
+          - "-c"
+          - python manage.py lms celery worker --loglevel=info --queues=edx.lms.core.high --hostname=edx.lms.core.default.%%h --concurrency=1
```
I am not familiar with this command, but: it seems that the hostname matches the default host, but you are always using the `edx.lms.core.high` queue (the same remark applies to other worker templates). Is that expected?

sampaccoud replied:
As discussed this morning, if you want to make this explicit, the right way to do it would be to declare the following settings explicitly in Arnold from Ansible variables: https://github.com/openfun/fun-platform/blob/master/config/lms/docker_run_production.py#L64

jmaupetit commented on apps/edxapp/templates/cms/dc_worker_low_priority.yml.j2:
```diff
@@ -0,0 +1,7 @@
+        command:
+          - "/bin/bash"
+          - "-c"
+          - python manage.py lms celery worker --loglevel=info --queues=edx.lms.core.high --hostname=edx.lms.core.low.%%h --concurrency=1
```
Those 3 templates should target the CMS not the LMS: `python manage.py cms [...] --queues=edx.cms.core.xx --hostname=edx.cms.core.low.%%h [...]`

jmaupetit commented on apps/edxapp/templates/lms/_configs/settings.yml.j2:
```diff
@@ -6,3 +6,8 @@ SITE_NAME: "{{ project_display_name }}"
+
+# Celery Broker
+CELERY_BROKER_TRANSPORT: redis
+CELERY_BROKER_HOSTNAME: redis
```
Should be `"{{ redis_app_host }}"`?

jmaupetit commented on apps/edxapp/templates/lms/dc_queue_high_priority.yml.j2:
```diff
@@ -0,0 +1,7 @@
+        command:
+          - "/bin/bash"
+          - "-c"
+          - python manage.py lms celery worker --loglevel=info --queues=edx.lms.core.high --hostname=edx.lms.core.high.%%h --concurrency=1
```
Same remark about targeted queues.

jmaupetit commented on apps/edxapp/templates/cms/dc.yml.j2:
```diff
@@ -10,9 +10,6 @@ metadata:
   replicas: 1  # number of pods we want
-  selector:
-    app: edxapp
-    deploymentconfig: edxapp-cms-{{ deployment_stamp }}
```
:+1:

jmaupetit committed:
👷(edxapp) test edxapp with redis enabled

Deploy Redis with edxapp to test Celery's default configuration and
workers deployment.

sampaccoud commented on apps/edxapp/vars/all/main.yml:
```diff
@@ -34,3 +34,12 @@ edxapp_nginx_image_name: "fundocker/openshift-nginx"
+edxapp_celery_lms_default_priority_queue: "edx.lms.core.default"
+edxapp_celery_lms_low_priority_queue: "edx.lms.core.low"
+edxapp_celery_lms_high_mem_queue: "edx.lms.core.high_mem"
+edxapp_celery_cms_high_priority_queue: "edx.cms.core.high"
```
maybe split in two groups lms and cms?

jmaupetit replied:
What do you mean by "group"?

sampaccoud replied:
add a comment line to split in two groups... just visual

sampaccoud commented on apps/edxapp/vars/all/main.yml:
```diff
@@ -34,3 +34,12 @@ edxapp_nginx_image_name: "fundocker/openshift-nginx"
+edxapp_celery_lms_high_mem_queue: "edx.lms.core.high_mem"
+edxapp_celery_cms_high_priority_queue: "edx.cms.core.high"
+edxapp_celery_cms_default_priority_queue: "edx.cms.core.default"
+edxapp_celery_cms_low_priority_queue: "edx.cms.core.low"
```
In fact I realized that edX is using the same queues for lms and cms. should we still split?

jmaupetit replied:
:thinking: I think we can also use the same queues. In this case, what about the workers? Should we keep two types of them?

sampaccoud replied:
so in fact edx IS using different queues. It is overriden in the aws.py setting file and I was looking at the common.py setting file. This being said, we could decide to run one worker per priority level and run both the lms and the cms queues on this worker... the setting when launching a worker is "queues" plural and can take several...

I'm for running 7 containers as it is simpler...

sampaccoud commented on tasks/manage_app.yml:
```diff
@@ -14,7 +14,7 @@
 - name: Set OpenShift objects to manage
   set_fact:
-    deployments: "{{ templates | map('regex_search', '.*/dc*.yml.j2$') | select('string') | list }}"
+    deployments: "{{ templates | map('regex_search', '.*/dc.*.yml.j2$') | select('string') | list }}"
```
will it take dc.yml.j2?

jmaupetit replied:
Yes `.*` means any occurrence (of any character) even none.

sampaccoud commented on tasks/manage_app.yml:
```diff
@@ -14,7 +14,7 @@
 - name: Set OpenShift objects to manage
   set_fact:
-    deployments: "{{ templates | map('regex_search', '.*/dc.yml.j2$') | select('string') | list }}"
+    deployments: "{{ templates | map('regex_search', '.*/dc.*.yml.j2$') | select('string') | list }}"
```
will this find dc.yml.j2?

sampaccoud replied:
Oh ok I just realized that this is a regex ! So you need to escape the other . no?
```
'.*/dc.*\.yml\.j2$'
```

jmaupetit replied:
Dot (`.`) works fine :wink:, but true!

jmaupetit committed:
🔊(playbook) display compiled template for create_object

Sometimes, displaying a compiled template before creating it eases its
debugging.

jmaupetit committed:
✨(apps) allow app declaration with no services

When we want to declare an application as deployable, without adding
other metadata, the services key was required. This is no longer the
case: an application (foo) can be activated as follows:

```yaml
apps:
  - name: foo
```

sampaccoud commented:
@jmaupetit Let's ship it! That's a great piece of team work. Thank you @rouja for kicking our ass on this one :rofl:
